node/actors/consensus/src/replica/state_machine.rs

```diff
@@ -96,38 +107,56 @@ impl StateMachine {
-            validator::ConsensusMsg::LeaderPrepare(_) => (
-                metrics::ConsensusMsgLabel::LeaderPrepare,
-                self.process_leader_prepare(ctx, consensus, signed_msg.cast().unwrap()),
-            ),
-            validator::ConsensusMsg::LeaderCommit(_) => (
-                metrics::ConsensusMsgLabel::LeaderCommit,
-                self.process_leader_commit(ctx, consensus, signed_msg.cast().unwrap()),
-            ),
+        let label = match &signed_msg.msg {
+            validator::ConsensusMsg::LeaderPrepare(_) => {
+                let res =
+                    match self.process_leader_prepare(ctx, consensus, signed_msg.cast().unwrap()) {
+                        Err(super::leader_prepare::Error::Internal(err)) => {
+                            return Err(err).context("process_leader_prepare()")
+                        }
+                        Err(err) => {
+                            tracing::warn!("process_leader_prepare(): {err:#}");
+                            Err(())
+                        }
+                        Ok(()) => Ok(()),
+                    };
+                metrics::ConsensusMsgLabel::LeaderPrepare.with_result(&res)
+            }
+            validator::ConsensusMsg::LeaderCommit(_) => {
+                let res =
+                    match self.process_leader_commit(ctx, consensus, signed_msg.cast().unwrap()) {
+                        Err(super::leader_commit::Error::Internal(err)) => {
+                            return Err(err).context("process_leader_commit()")
+                        }
+                        Err(err) => {
+                            tracing::warn!("process_leader_commit(): {err:#}");
+                            Err(())
+                        }
+                        Ok(()) => Ok(()),
+                    };
+                metrics::ConsensusMsgLabel::LeaderCommit.with_result(&res)
+            }
             _ => unreachable!(),
         };
-        metrics::METRICS.replica_processing_latency[&label.with_result(&result)]
-            .observe_latency(ctx.now() - now);
-        match result {
-            Ok(()) => Ok(()),
-            Err(err @ Error::ReplicaStateSave(_)) => Err(err.into()),
-            Err(err) => {
-                // Other errors from processing inputs are recoverable, so we just log them.
-                tracing::warn!("{err}");
-                Ok(())
-            }
-        }
+        metrics::METRICS.replica_processing_latency[&label].observe_latency(ctx.now() - now);
+        Ok(())
     }
 
     /// Backups the replica state to disk.
     pub(crate) fn backup_state(&self, ctx: &ctx::Ctx) -> anyhow::Result<()> {
+        let mut proposals = vec![];
+        for (number, payloads) in &self.block_proposal_cache {
+            proposals.extend(payloads.values().map(|p| storage::Proposal {
+                number: *number,
+                payload: p.clone(),
+            }));
+        }
```

Comment:
Nit: You could probably use `flat_map()` and `collect()` on an iterator instead of imperatively extending `proposals`; not that it matters much.

Reply:
ack